src/services/review/RepoCheckerService.ts

```diff
@@ -0,0 +1,223 @@
+    // create a log file in the root dir of all repos, in case of multiple errors
+    const logFilePath = path.join(process.cwd(), `log.txt`)
+
+    const logFile = fs.createWriteStream(logFilePath, { flags: "a" })
+    forEach(errors, (error) => {
+      logFile.write(`${error}\n`)
+    })
+    logFile.end()
+  }
+
+  emailErrors(errors: RepoError[]): ResultAsync<undefined, Error> {
+    // todo: send email to admin@isomer.gov.sg
+    return errAsync(new Error("Not implemented"))
+  }
+
+  getAllMediaPath(dirPath: string): ResultAsync<Set<string>, SiteCheckerError> {
+    const filesRootDir = path.join(dirPath, "files")
+    const imagesRootDir = path.join(dirPath, "images")
+
+    return ResultAsync.fromPromise(
+      this.traverseDirectory(filesRootDir, dirPath),
+      (e) => new SiteCheckerError(`Failed to traverse directory: ${e}`)
+    ).andThen((assetFilesPath) =>
+      ResultAsync.fromPromise(
+        this.traverseDirectory(imagesRootDir, dirPath),
+        (e) => new SiteCheckerError(`Failed to traverse directory: ${e}`)
+      ).map((imagesPath) => new Set([...assetFilesPath, ...imagesPath]))
+    )
+  }
+
+  // todo: refactor to use neverthrow
+  async traverseDirectory(dir: string, relativePath: string) {
+    const filePaths = new Set<string>()
+    const files = await fs.promises.readdir(dir, {
+      recursive: true,
+      withFileTypes: true,
+    })
+
+    const promises = files.map(async (entry) => {
+      const res = path.resolve(relativePath, path.join(entry.path, entry.name))
+      if (!entry.isDirectory()) {
+        filePaths.add(decodeURIComponent(res.slice(relativePath.length)))
+      }
+    })
+    await Promise.all(promises)
+    return filePaths
+  }
+
+  // todo: refactor to use neverthrow
+  async getAllPermalinks(repoPath: string, mdFiles: Set<string>) {
+    const setOfAllPermalinks = new Set<string>()
+    const files = [...mdFiles]
+
+    const promises = files
+      .filter((file) => !file.startsWith(".") && file.endsWith(".md")) // should not have .git
+      .map(async (file) => {
+        const regex = /(permalink: ){1}(.*)/gm
+        const fileContent = await fs.promises.readFile(
+          path.join(repoPath, file),
+          "utf8"
+        )
+        // match regex against file
+        const matches = fileContent.match(regex)
+        // if match, add to set
+        matches?.forEach((match) => {
+          const permalink = match.split(": ")[1].trim()
+          // permalinks for posts are generated dynamically based on the file name, and thus wont have conflict
+          if (!file.includes("_posts") && setOfAllPermalinks.has(permalink)) {
+            this.persistLogs(repoPath, [
+              new SiteCheckerError(
+                `Duplicate permalink ${permalink} found in ${file}`
+              ),
+            ])
+          } else {
+            setOfAllPermalinks.add(permalink)
+          }
+        })
+      })
+
+    await Promise.all(promises)
```

Comment:
Just wanted to check - intentional use of `all` over `allSettled`? All settled will ensure all promises are fulfilled even if some are rejected

Reply:
yup, this is the case!

Reply:
No point short circuiting here, we want other possible successful runs to occur regardless of some error states since they are independent events!

Reply:
@kishore03109 wait shouldn't we be using allSettled per ur description?

> Promise.all() will reject immediately upon any of the input promises rejecting

> Promise.allSettled() will wait for all input promises to complete, regardless of whether or not one rejects.

Reply:
arh apologies thought this was referring to a different code. yes, this one has to be `.all()` since we need to parse all files to get a good checker result

Reply:
to be clearer: I thought the code `promise.all()` here was in the function call `checkAllRepos()` (this is introduced in a later pr)

here, for a sane check of the repo,  I would have to parse the entire repo because it would not be possible check if an internal link is valid without the enumeration through all the permalinks in a repo